### ISSUE TITLE
Bug - Dependency injection broken for multiple REST API controller and service files

### ISSUE THREAD

**Issue by jkglasbrenner** [usnistgov/dioptra#30]
The dependency injector is causing multiple unit tests to fail, for example see https://github.com/usnistgov/dioptra/runs/5758593988?check_suite_focus=true#step:8:56. All of the failed tests emit a similar stack trace,

<details>
<summary>Stack trace – <code>test_create in tests/unit/restapi/experiment/test_controller.py</code></summary>

```
________________________ ERROR at setup of test_create _________________________

dependency_injector = <injector.Injector object at 0x7fe5336d80d0>

    @pytest.fixture
    def experiment_service(dependency_injector) -> ExperimentService:
>       return dependency_injector.get(ExperimentService)

tests/unit/restapi/experiment/test_service.py:53: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:9[61](https://github.com/usnistgov/dioptra/runs/5758593988?check_suite_focus=true#step:8:61): in get
    result = scope_instance.get(interface, binding.provider).get(self)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:283: in get
    return injector.create_object(self._cls)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:988: in create_object
    reraise(e, CallError(instance, init_function, (), additional_kwargs, e, self._stack))
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:209: in reraise
    raise exception.with_traceback(tb)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:984: in create_object
    self.call_with_injection(init, self_=instance, kwargs=additional_kwargs)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:1020: in call_with_injection
    owner_key=self_.__class__ if self_ is not None else callable.__module__,
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:109: in wrapper
    return function(*args, **kwargs)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:1065: in args_to_inject
    instance: Any = self.get(interface)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:961: in get
    result = scope_instance.get(interface, binding.provider).get(self)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:283: in get
    return injector.create_object(self._cls)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:988: in create_object
    reraise(e, CallError(instance, init_function, (), additional_kwargs, e, self._stack))
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:209: in reraise
    raise exception.with_traceback(tb)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:984: in create_object
    self.call_with_injection(init, self_=instance, kwargs=additional_kwargs)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:1006: in call_with_injection
    bindings = get_bindings(callable)
.tox/pytest/lib/python3.7/site-packages/injector/__init__.py:1147: in get_bindings
    type_hints = get_type_hints(callable, include_extras=True)
.tox/pytest/lib/python3.7/site-packages/typing_extensions.py:1302: in get_type_hints
    hint = typing.get_type_hints(obj, globalns=globalns, localns=localns)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/typing.py:1013: in get_type_hints
    value = _eval_type(value, globalns, localns)
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/typing.py:2[63](https://github.com/usnistgov/dioptra/runs/5758593988?check_suite_focus=true#step:8:63): in _eval_type
    return t._evaluate(globalns, localns)
/opt/hostedtoolcache/Python/3.7.12/x[64](https://github.com/usnistgov/dioptra/runs/5758593988?check_suite_focus=true#step:8:64)/lib/python3.7/typing.py:4[67](https://github.com/usnistgov/dioptra/runs/5758593988?check_suite_focus=true#step:8:67): in _evaluate
    eval(self.__forward_code__, globalns, localns),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: unsupported operand type(s) for |: '_GenericAlias' and 'NoneType'

<string>:1: TypeError
```

</details>

The [same unit tests all passed on March 10, 2022](https://github.com/usnistgov/dioptra/runs/5502911891?check_suite_focus=true#step:8:14) and the REST API code has not been updated since then. This means that this breakage is due to an update to one of the REST API's dependencies. Below is a comparison of the packages before and after the breakage occurred.

<details>
<summary>Package versions sdiff</code></summary>

```
# Left: March 10, 2022 packages (no breakage)
# Right: March 30, 2022 packages (breakage)

alembic==1.7.6                                                | alembic==1.7.7
aniso8601==9.0.1                                                aniso8601==9.0.1
                                                              > async-timeout==4.0.2
attrs==21.4.0                                                   attrs==21.4.0
boto3==1.21.17                                                | boto3==1.21.29
botocore==1.24.17                                             | botocore==1.24.29
certifi==2021.10.8                                              certifi==2021.10.8
cffi==1.15.0                                                    cffi==1.15.0
charset-normalizer==2.0.12                                      charset-normalizer==2.0.12
click==7.1.2                                                    click==7.1.2
cloudpickle==2.0.0                                              cloudpickle==2.0.0
coverage==6.3.2                                                 coverage==6.3.2
croniter==1.3.4                                                 croniter==1.3.4
cryptography==3.4.8                                             cryptography==3.4.8
cycler==0.11.0                                                  cycler==0.11.0
dask==2022.2.1                                                | dask==2022.2.0
databricks-cli==0.16.4                                          databricks-cli==0.16.4
Deprecated==1.2.13                                              Deprecated==1.2.13
distributed==2022.2.1                                         | distributed==2022.2.0
docker==5.0.3                                                   docker==5.0.3
entrypoints==0.4                                                entrypoints==0.4
Flask==1.1.4                                                    Flask==1.1.4
flask-accepts==0.18.4                                           flask-accepts==0.18.4
Flask-Injector==0.13.0                                          Flask-Injector==0.13.0
Flask-Migrate==2.7.0                                            Flask-Migrate==2.7.0
flask-restx==0.5.1                                              flask-restx==0.5.1
Flask-SQLAlchemy==2.5.1                                         Flask-SQLAlchemy==2.5.1
Flask-WTF==0.15.1                                               Flask-WTF==0.15.1
fonttools==4.30.0                                             | fonttools==4.31.2
freezegun==1.2.0                                              | freezegun==1.2.1
fsspec==2022.2.0                                                fsspec==2022.2.0
gitdb==4.0.9                                                    gitdb==4.0.9
GitPython==3.1.27                                               GitPython==3.1.27
greenlet==1.1.2                                                 greenlet==1.1.2
gunicorn==20.1.0                                                gunicorn==20.1.0
HeapDict==1.0.1                                                 HeapDict==1.0.1
idna==3.3                                                       idna==3.3
imageio==2.16.1                                                 imageio==2.16.1
imgaug==0.4.0                                                   imgaug==0.4.0
importlib-metadata==4.11.2                                    | importlib-metadata==4.11.3
importlib-resources==5.4.0                                    | importlib-resources==5.6.0
iniconfig==1.1.1                                                iniconfig==1.1.1
injector==0.19.0                                                injector==0.19.0
itsdangerous==1.1.0                                             itsdangerous==1.1.0
Jinja2==2.11.3                                                  Jinja2==2.11.3
jmespath==0.10.0                                              | jmespath==1.0.0
jsonschema==4.4.0                                               jsonschema==4.4.0
kiwisolver==1.3.2                                             | kiwisolver==1.4.2
locket==0.2.1                                                   locket==0.2.1
Mako==1.2.0                                                     Mako==1.2.0
MarkupSafe==2.0.1                                               MarkupSafe==2.0.1
marshmallow==3.14.1                                           | marshmallow==3.15.0
marshmallow-oneofschema==3.0.1                                  marshmallow-oneofschema==3.0.1
matplotlib==3.5.1                                               matplotlib==3.5.1
mitre-securing-ai @ file:///home/runner/work/dioptra/dioptra/   mitre-securing-ai @ file:///home/runner/work/dioptra/dioptra/
mlflow==1.24.0                                                  mlflow==1.24.0
msgpack==1.0.3                                                  msgpack==1.0.3
multimethod==1.7                                                multimethod==1.7
mypy-extensions==0.4.3                                          mypy-extensions==0.4.3
networkx==2.7.1                                               | networkx==2.6.3
numpy==1.22.3                                                 | numpy==1.21.5
opencv-python==4.5.5.64                                         opencv-python==4.5.5.64
packaging==21.3                                                 packaging==21.3
pandas==1.4.1                                                 | pandas==1.3.5
partd==1.2.0                                                    partd==1.2.0
pendulum==2.1.2                                                 pendulum==2.1.2
Pillow==9.0.1                                                   Pillow==9.0.1
pluggy==1.0.0                                                   pluggy==1.0.0
prefect==1.0.0                                                  prefect==1.0.0
prometheus-client==0.13.1                                       prometheus-client==0.13.1
prometheus-flask-exporter==0.19.0                               prometheus-flask-exporter==0.19.0
protobuf==3.19.4                                                protobuf==3.19.4
psutil==5.9.0                                                   psutil==5.9.0
py==1.11.0                                                      py==1.11.0
pycparser==2.21                                                 pycparser==2.21
pyparsing==3.0.7                                                pyparsing==3.0.7
pyrsistent==0.18.1                                              pyrsistent==0.18.1
pytest==6.2.5                                                   pytest==6.2.5
pytest-cov==3.0.0                                               pytest-cov==3.0.0
pytest-datadir==1.3.1                                           pytest-datadir==1.3.1
python-box==5.4.1                                             | python-box==6.0.1
python-dateutil==2.8.2                                          python-dateutil==2.8.2
python-slugify==6.1.1                                           python-slugify==6.1.1
pytz==2021.3                                                  | pytz==2022.1
pytzdata==2020.1                                                pytzdata==2020.1
PyWavelets==1.2.0                                             | PyWavelets==1.3.0
PyYAML==6.0                                                     PyYAML==6.0
querystring-parser==1.2.4                                       querystring-parser==1.2.4
redis==4.1.4                                                  | redis==4.2.0
requests==2.27.1                                                requests==2.27.1
rq==1.10.1                                                      rq==1.10.1
s3transfer==0.5.2                                               s3transfer==0.5.2
scikit-image==0.19.2                                            scikit-image==0.19.2
scipy==1.8.0                                                  | scipy==1.7.3
Shapely==1.8.1.post1                                            Shapely==1.8.1.post1
six==1.16.0                                                     six==1.16.0
smmap==5.0.0                                                    smmap==5.0.0
sortedcontainers==2.4.0                                         sortedcontainers==2.4.0
SQLAlchemy==1.4.32                                              SQLAlchemy==1.4.32
sqlparse==0.4.2                                                 sqlparse==0.4.2
structlog==21.5.0                                               structlog==21.5.0
tabulate==0.8.9                                                 tabulate==0.8.9
tblib==1.7.0                                                    tblib==1.7.0
text-unidecode==1.3                                             text-unidecode==1.3
tifffile==2022.2.9                                            | tifffile==2021.11.2
toml==0.10.2                                                    toml==0.10.2
tomli==2.0.1                                                    tomli==2.0.1
toolz==0.11.2                                                   toolz==0.11.2
tornado==6.1                                                    tornado==6.1
typing_extensions==4.1.1                                        typing_extensions==4.1.1
urllib3==1.26.8                                               | urllib3==1.26.9
websocket-client==1.3.1                                       | websocket-client==1.3.2
Werkzeug==1.0.1                                                 Werkzeug==1.0.1
wrapt==1.14.0                                                   wrapt==1.14.0
WTForms==3.0.1                                                  WTForms==3.0.1
zict==2.1.0                                                     zict==2.1.0
zipp==3.7.0                                                     zipp==3.7.0
```

</details>

The error messages indicate that the problem has something to do with type annotations. The [injector](https://github.com/alecthomas/injector) library uses the type annotations in its framework, although the package wasn't updated after March 10, so it must be due to an update in some other package. I suspect that this breakage is likely due to packages preparing for [PEP563](https://peps.python.org/pep-0563/), which changes the way that type annotations are handled[^1].

[^1]: There's been a lot of discussion over PEP563 since it has the potential to break popular Python packages like [Pydantic](https://github.com/samuelcolvin/pydantic/) and [FastAPI](https://github.com/tiangolo/fastapi), or at least things harder for newcomers to understand. There is a good discussion about this on the Python discussion boards: <https://discuss.python.org/t/type-annotations-pep-649-and-pep-563/11363>.

**Comment by jkglasbrenner** [comment]
I ran an experiment where I downgraded the packages in the _Package versions sdiff_ to their March 10th value. The process was as follows:

1. Select a package and revert it to its version as of March 10th. Keep all other packages the same.
2. Run the `restapi` unit tests to determine if this resolved the issue.
3. If it didn't resolve the issue, that means that package wasn't responsible, so undo the downgrade, select the next candidate package, and go back to step 1. If it did resolve the issue, then that means the package is responsible and the experiment is complete.

From this, I figured out that it was the [marshmallow](https://marshmallow.readthedocs.io/en/stable/) library's `3.14.1 -> 3.15.0` upgrade that caused the breakage, specifically marshmallow-code/marshmallow#1932. As I anticipated, this update was to prepare for [PEP 563](https://peps.python.org/pep-0563/). It also applies the syntax change to the Union type annotation [PEP 604](https://peps.python.org/pep-0604/), which seemed to be the specific issue.

**Comment by jkglasbrenner** [comment]
The reason this change to [marshmallow](https://marshmallow.readthedocs.io/en/stable/) broke [injector](https://github.com/alecthomas/injector) is because we didn't create explicit bindings or providers for all the classes that we inject. For example, there is no explicit binding or provider for the `ExperimentRegistrationFormSchema` type annotation in https://github.com/usnistgov/dioptra/blob/29ee50438a001715d1ca0f3038527fe286e74719/src/mitre/securingai/restapi/experiment/service.py#L48-L56. Injector's default behavior when there is no binding is to instantiate the class using the default values for its constructor. It will also check if the constructor has type annotations that are bound and require injection, and it was this check that failed as it didn't understand how to handle [PEP 604](https://peps.python.org/pep-0604/) changes in [marshmallow like the following](https://github.com/marshmallow-code/marshmallow/commit/6c4a42395fe42e0fba4b99b17ab66b5845088421?diff=unified#diff-0cddc4aac0e4e16946830a5807c264d65b7f14a2a1b1cb23b03028a36df3681aL1660-R1655),

```diff
-        schemes: typing.Optional[types.StrSequenceOrSet] = None,
+        schemes: types.StrSequenceOrSet | None = None,
```

Based on the stack trace message of `TypeError: unsupported operand type(s) for |: '_GenericAlias' and 'NoneType'`, the error occurs when injector calls [typing.get_type_hints](https://docs.python.org/3/library/typing.html#typing.get_type_hints) on a class using a type annotation like `typing.StrSequenceOrSet | None`. [typing.get_type_hints](https://docs.python.org/3/library/typing.html#typing.get_type_hints) doesn't understand how to evaluate this usage of the `|` operator, and the error is raised.

This is quite a subtle and interesting bug, as it emerges due to the interaction of code changes in two libraries in anticipation of [PEP 563](https://peps.python.org/pep-0563/) and [PEP 604](https://peps.python.org/pep-0604/). Each individual library behaves appropriately and as expected on its own. The real issue is that we didn't explicitly declare our injector bindings in all cases, and so we were relying on the default behavior [demonstrated in these examples in injector's README](https://github.com/alecthomas/injector#a-quick-example).

**Comment by jkglasbrenner** [comment]
To fix this, we need to declare explicit injector providers for all classes inheriting from marshmallow's `Schema` class that are injected elsewhere. For example, we would need to make the following change in [src/mitre/securingai/restapi/experiment/dependencies.py](https://github.com/usnistgov/dioptra/blob/1645f5b955ea99fa25b23b8316cf614a0ad2262c/src/mitre/securingai/restapi/experiment/dependencies.py),

```diff
# ACCESS THE FULL CC BY 4.0 LICENSE HERE:
# https://creativecommons.org/licenses/by/4.0/legalcode
"""Binding configurations to shared services using dependency injection."""
+from __future__ import annotations

from typing import Any, Callable, List

-from injector import Binder
+from flask_injector import request
+from injector import Binder, Module, provider
+
+from .schema import ExperimentRegistrationFormSchema
+
+
+class ExperimentRegistrationFormSchemaModule(Module):
+    @request
+    @provider
+    def provide_experiment_registration_form_schema_module(
+        self,
+    ) -> ExperimentRegistrationFormSchema:
+        return ExperimentRegistrationFormSchema()


def bind_dependencies(binder: Binder) -> None:
@@ -37,4 +50,4 @@ def register_providers(modules: List[Callable[..., Any]]) -> None:
        modules: A list of callables used for configuring the dependency injection
            environment.
    """
-    pass
+    modules.append(ExperimentRegistrationFormSchemaModule)
```

In addition, it's worth enforcing compliance with PEP 563 and PEP 604 by adding `from __future__ import annotations` to the top of all `restapi` files, since this is becoming the norm for several of Dioptra's dependencies.

This fix is implemented in #31, which passes all current tests.
